frame/support/procedural/src/pallet_error.rs

```diff
@@ -0,0 +1,114 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use frame_support_procedural_tools::generate_crate_access_2018;
+
+// Derive `PalletError`
+pub fn derive_pallet_error(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+	let syn::DeriveInput { ident: name, generics, data, .. } = match syn::parse(input) {
+		Ok(input) => input,
+		Err(e) => return e.to_compile_error().into(),
+	};
+
+	let frame_support = match generate_crate_access_2018("frame-support") {
+		Ok(c) => c,
+		Err(e) => return e.into_compile_error().into(),
+	};
+	let frame_support = &frame_support;
+	let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
+	let max_encoded_size = match data {
+		syn::Data::Struct(syn::DataStruct { fields, .. }) => match fields {
+			syn::Fields::Named(f) => {
+				let field_tys = f.named.iter().map(|field| &field.ty);
+				quote::quote! {
+					0_usize
+					#(.saturating_add(<
+						#field_tys as #frame_support::traits::PalletError
+					>::MAX_ENCODED_SIZE))*
+				}
+			},
+			syn::Fields::Unnamed(f) => {
+				let field_tys = f.unnamed.iter().map(|field| &field.ty);
+				quote::quote! {
+					0_usize
+					#(.saturating_add(<
+						#field_tys as #frame_support::traits::PalletError
+					>::MAX_ENCODED_SIZE))*
+				}
+			},
+			syn::Fields::Unit => quote::quote!(0),
+		},
+		syn::Data::Enum(syn::DataEnum { variants, .. }) => {
+			let field_tys = variants
+				.iter()
+				.map(|variant| match &variant.fields {
+					syn::Fields::Named(f) =>
+						Ok(Some(f.named.iter().map(|field| &field.ty).collect::<Vec<_>>())),
+					syn::Fields::Unnamed(f) =>
+						Ok(Some(f.unnamed.iter().map(|field| &field.ty).collect::<Vec<_>>())),
+					syn::Fields::Unit => Ok(None),
+				})
+				.collect::<Result<Vec<Option<Vec<&syn::Type>>>, syn::Error>>();
+
+			let field_tys = match field_tys {
+				Ok(tys) => tys.into_iter().flatten().collect::<Vec<_>>(),
+				Err(e) => return e.to_compile_error().into(),
+			};
+
+			// We start with `1`, because the discriminant of an enum is stored as u8
+			if field_tys.is_empty() {
+				quote::quote!(1)
+			} else {
+				let variant_sizes = field_tys.into_iter().map(|variant_field_tys| {
+					quote::quote! {
+						1_usize
+						#(.saturating_add(<
+							#variant_field_tys as #frame_support::traits::PalletError
+						>::MAX_ENCODED_SIZE))*
+					}
+				});
+
+				quote::quote! {{
+					let mut size = 1_usize;
+					let mut tmp = 0_usize;
+					#(
+						tmp = #variant_sizes;
+						size = if tmp > size { tmp } else { size };
+						tmp = 0_usize;
+					)*
+					size
+				}}
+			}
+		},
+		syn::Data::Union(syn::DataUnion { union_token, .. }) => {
+			let msg = "Cannot derive `PalletError` for union; please implement it directly";
+			return syn::Error::new(union_token.span, msg).into_compile_error().into()
+		},
+	};
```

Comment:
> take into account the attribute of codec like `skip` and `compact` and `encode_as`

That is a good point!



> of compact: compact u8 can take 2 bytes for example.

`MAX_ENCODED_SIZE` for this should then be `2`. Where is the problem?

Reply:
> > of compact: compact u8 can take 2 bytes for example.
> 
> `MAX_ENCODED_SIZE` for this should then be `2`. Where is the problem?

yes no problem I wanted to point that `Compact<T>` can have a maximum encoded size bigger than `T` in some cases

Reply:
Huh, this is a bit tricky, because if I understand it correctly, if I register an attribute helper like `codec` for this macro, then the attribute gets consumed by it, making it impossible for other derive macros like `Encode` to read the attribute and thus create a compact representation for the field.

I'm not 100% on this, would like to verify this.

Reply:
Sorry, looks like I'm wrong according to https://doc.rust-lang.org/reference/procedural-macros.html#derive-macro-helper-attributes:

> These attributes are [inert](https://doc.rust-lang.org/reference/attributes.html#active-and-inert-attributes), and their only purpose is to be fed into the derive macro that defined them. That said, they can be seen by all macros.